File: lecture14/exercise14-B/solution.py
#!/usr/bin/env python3

# Exercise 14-B: Max Min

import sys

# Functions

def read_numbers():
    try:
        n = int(sys.stdin.readline())
        k = int(sys.stdin.readline())
        v = [int(sys.stdin.readline()) for _ in range(n)]
    except ValueError:
        return 0, 0, None

    return n, k, v

def compute_unfairness(n, k, v):
    v.sort()                    # Order from smallest to largest
    u = v[n - 1] - v[0]         # Initial unfairness
    for i in range(0, n - k + 1):
        d = v[i + k - 1] - v[i] # Difference between max and min in sequence
        u = min(u, d)           # Take smallest unfairness
    return u

# Main execution

def main():
    n, k, v = read_numbers()
    while n: 
        print(compute_unfairness(n, k, v))
        n, k, v = read_numbers()

if __name__ == '__main__':
    main()

File: lecture14/exercise14-B/solution.cpp
// Exercise 14-B: Max Min

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// Functions

bool read_numbers(int &n, int &k, vector<int> &v) {
    int t;

    if (!(cin >> n >> k)) {
    	return false;
    }

    v.clear();
    for (int i = 0; i < n; i++) {
    	cin >> t;
    	v.push_back(t);
    }

    return true;
}

int compute_unfairness(int n, int k, vector<int> &v) {
    int u, d;

    sort(v.begin(), v.end());

    u = v[n - 1] - v[0];		// Initial unfairness
    for (int i = 0; i <= (n - k); i++) {
	d = v[i + k - 1] - v[i];	// Difference
	u = min(d, u);			// Update unfairness
    }

    return u;
}

// Main execution

int main(int argc, char *argv[]) {
    int n, k;
    vector<int> v;

    while (read_numbers(n, k, v)) {
	cout << compute_unfairness(n, k, v) << endl;
    }

    return 0;
}
